src/workerd/io/trace.h

```diff
@@ -519,4 +311,190 @@ inline kj::String truncateScriptId(kj::StringPtr id) {
+
+private:
+  kj::Maybe<kj::Own<SpanObserver>> observer;
+
+  friend class SpanBuilder;
+
+  SpanParent(kj::Maybe<kj::Own<SpanObserver>> observer): observer(kj::mv(observer)) {}
+};
+
+class SpanBuilder {
+  // Interface for writing a span. Essentially, this is a mutable interface to a `Span` object,
+  // given only to the code which is meant to create the span, whereas code that merely collects
+  // and reports spans gets the `Span` type.
+  //
+  // The reason we use a separate builder type rather than rely on constness is so that the methods
+  // can be no-ops when there is no observer, avoiding unnecessary allocations. To allow for this,
+  // SpanBuilder is designed to be write-only -- you cannot read back the content. Only the
+  // observer (if their is one) receives the content.
+
+public:
+  explicit SpanBuilder(kj::Maybe<kj::Own<SpanObserver>> observer, kj::StringPtr operationName)
+      { KJ_IF_MAYBE(o, observer) { state.emplace(kj::mv(*o), operationName); } }
+  // Create a new top-level span that will report to the given observer. If the observer is null,
+  // no data is collected.
+  //
+  // `operationName` should be a string literal with infinite lifetime.
+
+  SpanBuilder(SpanBuilder&& other) = default;
+  SpanBuilder& operator=(SpanBuilder&& other) = default;
+  KJ_DISALLOW_COPY(SpanBuilder);
+
+  ~SpanBuilder() noexcept(false);
+
+  void setStartTime(kj::Date time);
+  // Set a different start time. This is optional; by default the time when the builder was
+  // constructed will be used.
+
+  using TagValue = Span::TagValue;
+  void setTag(kj::StringPtr key, TagValue value);
+  // `key` must point to memory that will remain valid all the way until this span's data is
+  // serialized.
+
+  void addLog(kj::Date timestamp, kj::StringPtr key, TagValue value);
+  // `key` must point to memory that will remain valid all the way until this span's data is
+  // serialized.
+  //
+  // The differences between this and `setTag()` is that logs are timestamped and may have
+  // duplicate keys.
+
+private:
+  struct State {
+    kj::Own<SpanObserver> observer;
+    // The observer, or null if not being observed.
+
+    Span span;
+    // The under-construction span.
+
+    State(kj::Own<SpanObserver> observer, kj::StringPtr operationName)
+        : observer(kj::mv(observer)), span(operationName) {}
+  };
+  kj::Maybe<State> state;
+
+  friend class SpanParent;
+};
+
+class SpanObserver: public kj::Refcounted {
+  // Abstract interface for observing trace spans reported by the runtime. Different
+  // implementations might support different tracing back-ends, e.g. Trace Workers, Jaeger, or
+  // whatever infrastrure you prefer to use for this.
+  //
+  // A new SpanObserver is created at the start of each Span. The observer is used to report the
+  // span data at the end of the span, as well as to construct child observers.
+
+public:
+  virtual kj::Own<SpanObserver> newChild() = 0;
```

Comment:
Perhaps something we might do in the future is give the observer the ability to provide the time for spans for construction and destruction, which would be handy for putting spans in trace workers.

Reply:
Not quite sure what you mean, can you reword?

Reply:
Rather than hard-code using the system clock for start and end times, the observer could provide the clock instead. It might come in handy in the future for a public span tracing API that's part of trace workers.